src/navigation_controller.ts

```diff
@@ -148,33 +159,65 @@ export class NavigationController {
   }
 
   /**
-   * Sets whether the navigation controller has focus. This will enable keyboard
-   * navigation if focus is now gained. Additionally, the cursor may be reset if
-   * it hasn't already been positioned in the workspace.
+   * Sets whether the navigation controller has toolbox or flyout focus. This
+   * will enable keyboard navigation in the toolbox or flyout.
    *
-   * @param workspace the workspace that now has input focus.
+   * @param workspace the workspace that now has toolbox/flyout input focus.
    * @param isFocused whether the environment has browser focus.
    */
-  setHasFocus(workspace: WorkspaceSvg, isFocused: boolean) {
-    this.hasNavigationFocus = isFocused;
+  updateToolboxOrFlyoutFocus(workspace: WorkspaceSvg, isFocused: boolean) {
     if (isFocused) {
-      this.navigation.focusWorkspace(workspace, true);
+      if (!workspace.getToolbox()) {
+        this.navigation.focusFlyout(workspace);
+      } else {
+        this.navigation.focusToolbox(workspace);
+      }
+      this.navigationFocus = NAVIGATION_FOCUS_MODE.TOOLBOX_OR_FLYOUT;
+    } else {
+      this.navigation.blurToolboxAndFlyout(workspace);
+      this.navigationFocus = NAVIGATION_FOCUS_MODE.NONE;
     }
   }
 
+  /**
+   * Sets whether the navigation controller has workspace focus. This will
+   * enable keyboard navigation within the workspace. Additionally, the cursor
+   * may be reset if it hasn't already been positioned in the workspace.
+   *
+   * @param workspace the workspace that now has workspace input focus.
+   * @param isFocused whether the environment has browser focus.
+   */
+  updateWorkspaceFocus(workspace: WorkspaceSvg, isFocused: boolean) {
+    if (isFocused) {
+      this.navigation.focusWorkspace(workspace, true);
+      this.navigationFocus = NAVIGATION_FOCUS_MODE.WORKSPACE;
+    } else this.navigationFocus = NAVIGATION_FOCUS_MODE.NONE;
```

Comment:
Style: Please use brackets even for a one-liner. We only skip them for `if (x) return;`

Reply:
Ah, I was unsure on this case. :) Fixed.